### ISSUE TITLE
Recommendation, open questions

### ISSUE THREAD

**Issue by berndl** [anno4j/anno4j#26]
Looked at Apache Mahout - Taste (https://mahout.apache.org/users/recommender/recommender-documentation.html):
- Mainly focused on collaborative filtering, item-based recommendations are possible but only in a very restrictive way. They claim that items are rather static, so their similarity can be precomputed.
- Their basic engine needs to be supported with a list of precomputed item-similarities.
- Maybe workflow: Given an annotation, define requested similarities -> Server queries related annotations (referred from its type) and calculates similarities -> Feed to taste -> Generate Recommendations

-> Questions:
- Should our annotation similarity be static, calculated from time to time?
- Should they be dynamic, calculated at request-time?
- Between what kind of annotations/annotation types does one want to have similarities? All of them?
- Cross-annotation similarity: how to define those similarities?
- Can the desired similarity be defined on recommendation-request time?
- Can the algorithms from taste be used on lower level annotation features?
- Do we need the support of a framework altogether? Recommendations or similarities could be implemented at annotation level..

-> Ideas:
- Basic annotation similarity based on annotation level and target level (e.g. User, Agent, Target resource, same selector/fragment, etc.)
- Specific annotation similarity: Defined by some rules?

**Comment by mgrani** [comment]
Collaborative Filtering can be very tricky in the context of annotations. Namely because you have a potential very huge set of items that are annotated. So i would rather see a content based approach, i.e. similarity among the content of annotation and the content of an item, rather than using collaborative filtering. 

In order to set a particular focus, i would assume that the recommendation service provides means for querying annotations and allows for setting specific user preferences (e.g. the location of a person, the language, the age etc.). User preferences should be taken into account, but dont form a hard constraint on the query.

The hardest part will be the cross media part, which requires some thought in the definition. 

To become more clear on the task, i would  assume we have an item $I$ that consists of several media parts organised in the form of a document object model $DOM(I)$. A simple example would be a web page. Further, we have a set of Annotations $A$ which are linked to each other and to an part of an Item $I$. 

Given a new item $I_n$ we have two cases:
1. $I_n$ already has annotations associated with its parts $DOM (I)$ and hence we need to filter those annotations based on the user perferences. Similarity between annotations might be also taken into account in order to increase the number of relevant annotations. 
2.   $I_n$ does not have annotations associated with its parts  $DOM (I)$. If so, we can either find similar items to $I_n$ and then use the associations there or we try to extract the semantic of the content from $I_n$ or parts of it and then find similar annotations. The later part could be done by the [EEXCESS C4 Plugin](https://github.com/EEXCESS/c4) currently under development. In the cross media case this could be challenging.

Overall, i would opt for a query based system that provides 1 or 2 via an sophisticated, but limited query scheme. So you need to work on similarity schemes between items/part of items and annotations, annotations and annotations and items and items.

**Comment by berndl** [comment]
I totally agree with you, that collaborative filtering should not be focused at the start (although it seems, that collaborative filtering seems be more in the focus point of the current recommendation scene). There would be another level of user and user interaction necessary, which is not supported by the plain annotations and their items.

The cross-media part indeed will be tricky. I suppose we will focus basic annotation recommendation at first, as there is some information that all annotations share, and combine that with, as you mentioned, first simple query operations. Beyond that, combining annotations with different motivations and contents will be hard. Although my point of view of annotations might be to clouded from MICO use-cases here ;)

Your idea of seeing an annotation not just solely by itself, but rather in its context of the whole media item is interesting. That might open more possibilities to recommend certain things. And I am very looking forward to combine Anno4j with your C4 plugin!

I was currently looking for ways of expressing the similarity between annotations. Here as well, many things are supported for collaborative filtering, content based seemed a little more thin (at a first glance). Hope to support more insights (and maybe code ;) ) soon.

**Comment by berndl** [comment]
- Allow SKOS predicates to indicate similarity, next to the numerical (standard) similarity?
- What is the target of a similarity annotation? Currently we point the target of the annotation to its subject, as it is the "reason" why we calculate a similarity at all.

**Comment by berndl** [comment]
Basic similarities between annotations can now be generated, by specifying the two annotations (subject and object) and defining an algorithm to calculate the similarity.
This creates a similarity annotation, which makes use of RDF reification in order to express that a given annotation is similar (to a certain degree) to another annotation.

Querying (e.g. "give me annotations similar to my given one") will follow.
